### ISSUE TITLE
The reference manual should document which environment variables are necessary to get a working coq

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#17158]
What environment variables need to be appropriately set for `coqc` to work?  What environment variables can be set to impact the functioning of Coq?  https://github.com/coq/coq/issues/16334 suggests `OCAMLFIND_CONF` (though this is not listed by `coqc -config`)

And which variable will fix the error message `could not find META.coq-core.`?  (See also [here](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/could.20not.20find.20META.2Ecoq-core/near/323389514))

**Comment by silene** [comment]
No variable will ever fix the error message `could not find META.coq-core`. The Findlib configuration contains absolute paths. So, even if you set `OCAMLFIND_CONF` so that Findlib locate its configuration in its new location, the paths it contains still point to the old location.

**Comment by ejgallego** [comment]
> What environment variables need to be appropriately set for coqc to work?

If you are using plugins, you need the findlib to be properly configured, the set of variables is specific to findlib and we should reference their documentation, not duplicate it.

> What environment variables can be set to impact the functioning of Coq?

In addition to the ones that OCaml and findlib (and mabye zarith?) understand, `git grep getenv` says:
#### Configure + build

- all of Dune vars
- `COQ_CONFIGURE_PREFIX` `PATH`

#### Coq (some may not be used, or only used by CoqIDE)

- `HOMEDRIVE`, `HOMEPATH`, `XDG_DATA_DIRS`, `PATH`, `SSRMATCHINGDEBUG`

#### CoqIDE

- `PATH`, `GTK_CSD`, `HOME` (in OSX), `OCAMLRUNPARAM`, `OCAML_GC_STATS`, `OCAMLPATH`, `COQWORKMGR_SOCK`, `COQWORKMGR_SOCK`, `TERM`, `COQ_COLORS`

> And which variable will fix the error message could not find META.coq-core.?

That could be a Coq bug (as we have some horrible hacks to integrate with findlib)

Hard to know without a precise way to reproduce.

**Comment by JasonGross** [comment]
> No variable will ever fix the error message `could not find META.coq-core`. The Findlib configuration contains absolute paths. So, even if you set `OCAMLFIND_CONF` so that Findlib can locate its configuration in its new location, the paths it contains still point to the old location.

This seems horrible for modularity.  This means that it's impossible to have two parallel versions of Coq coming from CI artifacts which built Coq using different versions of OCaml.

**Comment by ejgallego** [comment]
Note as said in Zulip that this error message doesn't come from findlib but from some hacks Coq is doing and should be removed.

> This means that it's impossible to have two parallel versions of Coq coming from CI artifacts which built Coq using different versions of OCaml.

I'm not sure I follow, how is that impossible?

**Comment by JasonGross** [comment]
Ah, I was assuming the error was coming from findlib, and also partially responding to
> And by the way, Coq works just fine even when `OCAMLFIND_CONF` is not set, as long as you did not remove the Findlib files from their expected location. In other words, this bug report is the same as randomly deleting files and then complaining that a tool no longer works.
> _Originally posted by @silene in https://github.com/coq/coq/issues/16334#issuecomment-1403152916_

Insofar as the Coq only works when you have the version of OCaml/findlib lying around *that you built Coq with* and that version *must be in the same directory as it was when Coq was built*, you cannot run two different `coqc` processes simultaneously that are expecting to find in the same location two different versions of OCaml/findlib.

**Comment by ejgallego** [comment]
> Ah, I was assuming the error was coming from findlib, and also partially responding to

I'm afraid this seems like a Coq bug, that code needs more work, there are few open issues.

> must be in the same directory as it was when Coq was built

Umm, actually for runtime Coq only depends on the findlib metadata, which as far as I know can be configured to relocate. I believe Coq platform does that. It is not really supported but that works, IIUC, as long as the metadata versioning doesn't break up.

But it is pretty stable.

What needs to be identical to build time is the plugins.

I'm sorry this gets so confusing for users, let me try the steps to relocate stuff.

1. First you want to be sure vanilla Coq works fine, this is done with `coqtop -boot -noinit`
2. The second step is having Coq locate the standard library, that's done with a set of heuristics, setting `COQLIB` to where `Prelude.vo` is should suffice (there are more methods)
3. Now Coq can find the stdlib, so `coqtop -noinit` shoud work warning-free. However, .cmxs files for the plugins need findlib so the prelude can be loaded. For this, findlib needs a configuration file (tho I think it can work without one).

Once the findlib config is set, everything should work (TM)!

**Comment by silene** [comment]
> I'm afraid this seems like a Coq bug

Coq calls `Findlib.package_meta_file package`; Findlib reads the configuration file located at `OCAMLFIND_CONF`; this configuration file refers to obsolete directories that no longer contain any META files; Findlib answers it did not find any file; Coq raises an error. This is neither a Coq bug nor a Findlib bug.

**Comment by JasonGross** [comment]
> this configuration file refers to obsolete directories that no longer contain any META files

Coq modifies the configuration file at install time?  (Note that the findlib installation is unchanged here, unlike in #16334: here I've simply relocated the Coq install tree)

> This is neither a Coq bug nor a Findlib bug.

Even if it is not a bug, it is a regression and a misfeature that hampers usability.

**Comment by silene** [comment]
> to have two parallel versions of Coq coming from CI artifacts which built Coq using different versions of OCaml.

Are you calling `opam switch foo` and `eval $(opam env)` when you switch from one version of OCaml to the other?

**Comment by JasonGross** [comment]
> > to have two parallel versions of Coq coming from CI artifacts which built Coq using different versions of OCaml.
> 
> Are you calling `opam switch foo` and `eval $(opam env)` when you switch from one version of OCaml to the other?

No.  (But this is all hypothetical; I think the error message in this issue arises even when only one version of OCaml was used.). Is there a way to extract from coqc or from the CI build artifact what opam switch I should be using?

But this is also impossible if I ever want to run both versions of Coq simultaneously (to speed up minimization, for example), because AFAIK, `opam switch` is a global operation.

**Comment by silene** [comment]
Indeed, `opam switch` is a global operation, but it only matters if you are calling an `opam` command. On the other hand, `eval $(opam env)` is a local operation, by definition of `eval`. You can combine both into `eval $(opam env --switch=foo)`, if needed.

Running several Coq versions in parallel is something I do on a daily basis. I never have to mess with environment variables past `opam env`.

**Comment by silene** [comment]
> Is there a way to extract from coqc or from the CI build artifact what opam switch I should be using?

If you are able to run `coqtop -config`, then the Opam switch is apparent from the output, e.g., from the `OCAMLFIND` line.

**Comment by ejgallego** [comment]
Actually I think I got confused and @silene is right, this not a Coq bug. In your case it seems you are asking for the findlib-only path, and that indeed fails because you didn't set findlib properly.

However your report is lacking many critical details so it is hard to know what is going on.

But I think that's the key point, you wrote:

> Oh, ew, is `Declare ML Module "coq-core.plugins.ltac"`. no longer something that I can just dump in a file loaded with -nois to get Ltac support?

I'm unsure what you mean by "no longer", but that's a new feature for 8.16. And indeed, it requires findlib. The legacy loader is still available as `Declare ML Module "ltac_plugin".`

> Even if it is not a bug, it is a regression and a misfeature that hampers usability.

What is exactly a regression? If you are using a new 8.16 features, there can't be a regression.

(By the way the title of the issue is "document what env vars are needed", so I guess you are talking about other the other issue)

Indeed, we can improve the documentation, but it seems there is no bug here. You are using a new 8.16 feature (`Declare ML` with findlib name) and you have not configured findlib properly. 

Configuring findlib seems confusing for users, but I think I wrote above a good base, let's repeat:
- `OCAMLFIND_CONF` has to point out to a place where the right `findlib.conf` is setup
- `findlib.conf` must point to the right `lib` directory where the OCaml packages are.

**Comment by JasonGross** [comment]
> However your report is lacking many critical details so it is hard to know what is going on.

Sorry, I'm not super-comfortable with docker and so am trying to guess what's causing the error message at https://github.com/coq-community/run-coq-bug-minimizer/actions/runs/4014049671/jobs/6894103859#step:5:49535 without having a local setup reproducing it.

The thing that I'm doing here is: running the docker image, downloading two artifacts of Coq that were built with the same ocaml switch that I'm on, unpacking one of them to `/github/workspace/builds/coq/coq-failing/{_install,_build}_ci/...` and the other to `/github/workspace/builds/coq/coq-passing/{_install,_build}_ci/...`.  When I invoke `coqc` directly on the file it seems to run fine (maybe?), but when I invoke it through `coq_makefile`* it gives the error I mentioned?

I was assuming that someone would have an easy explanation for what could cause `could not find META.coq-core.`, but none of the answers given so far explain why using the same docker image that Coq CI uses but unpacking the Coq artifact to a different directory would result in this error message.  (AIUI, Coq CI scripts are not modifying the findlib configuration, but are also not installing their files to system directories.  Unless part of running `make ci-foo` in Coq re-runs Coq's install script in a way that changes the global ocamlfind configuration?)


> Configuring findlib seems confusing for users, but I think I wrote above a good base, let's repeat:
> 
> * `OCAMLFIND_CONF` has to point out to a place where the right `findlib.conf` is setup
> * `findlib.conf` must point to the right `lib` directory where the OCaml packages are.

Can you point to the line in Coq's CI where this is set up?  As far as I can tell, files are installed under `$PWD/_install_ci/lib` not `/root/.opamcache/*/lib`, so they're not being installed in the right place for findlib to pick them up automatically.  The only plausibly relevant line I can find is
https://github.com/coq/coq/blob/c647ce5f3151203a2867fe21bfe29cad03e6e38d/dev/ci/ci-common.sh#L15
but you claim that `OCAMLPATH` is not used by `coqc` at runtime except in CoqIDE.

> The legacy loader is still available as `Declare ML Module "ltac_plugin".`

Ah, I was under the impression that this was going to disappear at some point.  Is it planned to be kept around forever?

<details><summary>*<code>coq_makefile</code> invocation</summary>

```
/github/workspace/builds/coq/coq-passing/_install_ci/bin/coq_makefile COQC = /github/workspace/builds/coq/coq-passing/_install_ci/bin/coqc.orig COQDEP = /github/workspace/builds/coq/coq-passing/_install_ci/bin/coqdep -o Makefile30yek72s.coq -R /github/workspace/builds/coq/coq-passing/_build_ci/metacoq/template-coq/theories MetaCoq.Template -R /github/workspace/builds/coq/coq-passing/_install_ci/lib/coq/theories Coq -Q /github/workspace/cwd Top -Q /github/workspace/builds/coq/coq-passing/_install_ci/lib/coq/user-contrib/Equations Equations -Q /github/workspace/builds/coq/coq-passing/_install_ci/lib/coq/user-contrib/Ltac2 Ltac2 -Q /github/workspace/builds/coq/coq-passing/_install_ci/lib/coq/user-contrib/MetaCoq MetaCoq -I /github/workspace/builds/coq/coq-passing/_build_ci/metacoq/template-coq/src -I /github/workspace/builds/coq/coq-passing/_build_ci/metacoq/template-coq -R /github/workspace/builds/coq/coq-passing/_build_ci/metacoq/template-coq/theories MetaCoq.Template -R /github/workspace/builds/coq/coq-passing/_install_ci/lib/coq/theories Coq -Q /github/workspace/cwd Top -Q /github/workspace/builds/coq/coq-passing/_install_ci/lib/coq/user-contrib/Equations Equations -Q /github/workspace/builds/coq/coq-passing/_install_ci/lib/coq/user-contrib/Ltac2 Ltac2 -Q /github/workspace/builds/coq/coq-passing/_install_ci/lib/coq/user-contrib/MetaCoq MetaCoq -I /github/workspace/builds/coq/coq-passing/_build_ci/metacoq/template-coq/src -I /github/workspace/builds/coq/coq-passing/_build_ci/metacoq/template-coq -arg -q -arg -w -arg -deprecated-native-compiler-option -arg -native-compiler -arg no bug_01.v
make -k -f Makefile30yek72s.coq KEEP_ERROR=1 bug_01.glob
```
</details>

**Comment by JasonGross** [comment]
> Indeed, `opam switch` is a global operation, but it only matters if you are calling an `opam` command. On the other hand, `eval $(opam env)` is a local operation, by definition of `eval`. You can combine both into `eval $(opam env --switch=foo)`, if needed.

Neat, I was not aware of this, thanks!

> > Is there a way to extract from coqc or from the CI build artifact what opam switch I should be using?
> 
> If you are able to run `coqtop -config`, then the Opam switch is apparent from the output, e.g., from the `OCAMLFIND` line.

Neat, thanks!


> Running several Coq versions in parallel is something I do on a daily basis. I never have to mess with environment variables past `opam env`.

@silene This sounds like you have a separate opam switch and a separate copy of findlib for each Coq installation.  Can this work if I built two versions of Coq on the same opam switch, and want them to run simultaneously?  (e.g., I got them from CI build artifacts, or I built them locally)

> Indeed, we can improve the documentation, but it seems there is no bug here at least in the way you phrase it.

I am confused about the state of things.  Insofar as `findlib.conf` must point to the right `lib` directory and there's no way to override it with environment variables, you cannot simultaneously run two different versions of coqc based on the same opam switch.  This was supported before the switch to findlib, and I'd like this to continue being supported for coqbot auto bug minimization.  However, insofar as "pointing at Coq's install location" does not require futzing with global state, my request is just that the necessary configuration be documented.

Said another way, I want an answer to the question: suppose I relocate my Coq install directory from the directory it was originally installed to; what environment variables do I need to set for things to keep working?  (Insofar as the answer to this is "changing environment variables isn't enough", I call that a regression.)  (And, ideally, there'd be a similar answer to relocating the OCaml installation.)  (This seems like it needs to exist for the Mac and Windows Coq Platform installers to work, right?)

**Comment by silene** [comment]
> you claim that `OCAMLPATH` is not used by `coqc` at runtime except in CoqIDE.

Yes and no. While Coq does not really care about `OCAMLPATH`, it still has a major impact. First, Findlib uses `OCAMLPATH` to locate packages and so does Coq transitively. Second, Coq extends `OCAMLPATH` with some additional directories before Findlib gets to read it. It is a bit hard to be exhaustive when it comes to what these additional directories are. It should be something like `COQPATH` and its subdirectories, `XDG_DATA_foo/coq` and their subdirectories, `COQCORELIB/..`, options `-I`, various directories set at configuration time, etc. (Most of those are to ensure compatibility with dubious setups. In an ideal world, Coq would not need to touch `OCAMLPATH`.)

**Comment by ejgallego** [comment]
Thanks @JasonGross for the extra details, I know what is going on now.

> but you claim that OCAMLPATH is not used by coqc at runtime except in CoqIDE.

As Guillaume says it is used by findlib, so indeed in your case, the opam switch and findlib config is fine.

What you need is indeed to tell findlib that extra ocaml libraries are in `/github/workspace/builds/coq/coq-failing/_install_ci/`, using `OCAMLPATH=github/workspace/builds/coq/coq-failing/_install_ci/lib`

I think that should fix your problem! (do the same for 

> Ah, I was under the impression that this was going to disappear at some point. Is it planned to be kept around forever?

We will remove it indeed, but we need to fix a bug in Coq first. What I meant is that technically that's not a regression in the sense that you only face this problems if you select the new loading method.

But the situation is not great, I agree.

**Comment by ejgallego** [comment]
Doc for ocamlpath is at http://projects.camlcity.org/projects/dl/findlib-1.4/doc/ref-html/r775.html
